frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -28,6 +28,45 @@ static const std::vector<const IR::IDeclaration *> empty;
 
 ResolutionContext::ResolutionContext() { anyOrder = P4CContext::get().options().isv1(); }
 
+std::vector<const IR::IDeclaration *> *ResolutionContext::memoizeDeclarations(
+    const IR::INamespace *ns) const {
+    auto decls = ns->getDeclarations();
+    if (auto nest = ns->to<IR::INestedNamespace>()) {
+        // boost::adaptors::reverse expects a reference
+        auto temp = nest->getNestedNamespaces();
+        for (auto nn : boost::adaptors::reverse(temp)) decls = nn->getDeclarations()->concat(decls);
+    }
+
+    return (namespaceDecls[ns] = decls->toVector());
+}
+
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclarations(
+    const IR::INamespace *ns) const {
+    auto nsIt = namespaceDecls.find(ns);
+    auto *declVec = nsIt != namespaceDecls.end() ? nsIt->second : memoizeDeclarations(ns);
+    return Util::Enumerator<const IR::IDeclaration *>::createEnumerator(*declVec);
```

Comment:
I suspect you could get a much bigger gain here by getting rid of the java-style `Util::Enumerator` (which does a bunch of dynamic allocation and generates much garbage to be collected by the GC) and using C++ style enumeration instead, but that qould require rewriting all the code that uses this (many passes).

One could think about having a second interface in a ResolutionContext that allows for thatm and then incrementally change other code to use the new interface.  But that is a bigger project,

Reply:
Created an issue #4412.